File: english-hub-frontend/src/services/exerciseService.ts
import { ExerciseCreateRequest, ExerciseResponse } from '@/types/exerciseType';
import axiosInstance from './axiosInstance';
import {
  QuestionCreateRequest,
  QuestionGroupResponse,
  QuestionResponse,
  QuestionUpdateRequest,
} from '@/types/questionType';

export async function getAllExercises(
  courseId: string
): Promise<ExerciseResponse[]> {
  const response = await axiosInstance.get(`/courses/${courseId}/exercises`);
  return response.data.result;
}

export async function getExerciseById(
  courseId: string,
  exerciseId: string
): Promise<ExerciseResponse> {
  const response = await axiosInstance.get(
    `/courses/${courseId}/exercises/${exerciseId}`
  );
  return response.data.result;
}

export async function createExercise(
  courseId: string,
  exercise: ExerciseCreateRequest
): Promise<ExerciseResponse> {
  const response = await axiosInstance.post(
    `/courses/${courseId}/exercises`,
    exercise
  );
  return response.data.result;
}

export async function updateExercise(
  courseId: string,
  exerciseId: string,
  exercise: ExerciseCreateRequest
): Promise<ExerciseResponse> {
  const response = await axiosInstance.put(
    `/courses/${courseId}/exercises/${exerciseId}`,
    exercise
  );
  return response.data.result;
}

export async function deleteExercise(
  courseId: string,
  exerciseId: string
): Promise<string> {
  const response = await axiosInstance.delete(
    `/courses/${courseId}/exercises/${exerciseId}`
  );
  return response.data.message;
}

export async function addQuestionToExercise(
  courseId: string,
  exerciseId: string,
  question: QuestionCreateRequest
): Promise<QuestionResponse> {
  const response = await axiosInstance.post(
    `/courses/${courseId}/exercises/${exerciseId}/question`,
    question
  );
  return response.data.result;
}

export async function updateQuestionInExercise(
  courseId: string,
  exerciseId: string,
  questionId: string,
  question: QuestionUpdateRequest
): Promise<QuestionResponse> {
  console.log(question);
  
  const response = await axiosInstance.put(
    `/courses/${courseId}/exercises/${exerciseId}/questions/${questionId}`,
    question
  );
  return response.data.result;
}

export async function addQuestionsToExercise(
  courseId: string,
  exerciseId: string,
  questions: QuestionCreateRequest[]
): Promise<QuestionResponse[]> {
  const response = await axiosInstance.post(
    `/courses/${courseId}/exercises/${exerciseId}/questions`,
    questions
  );
  return response.data.result;
}

export async function getQuestionsFromExercise(
  courseId: string,
  exerciseId: string
): Promise<QuestionResponse[]> {
  const response = await axiosInstance.get(
    `/courses/${courseId}/exercises/${exerciseId}/questions`
  );
  console.log(response.data.result);
  
  return response.data.result;
}

export async function getQuestionGroupsFromExercise(
  courseId: string,
  exerciseId: string
): Promise<QuestionGroupResponse[]> {
  const response = await axiosInstance.get(
    `/courses/${courseId}/exercises/${exerciseId}/questions/groups`
  );
  return response.data.result;
}

File: english-hub-backend/src/main/java/com/example/englishhubbackend/service/ExerciseService.java
package com.example.englishhubbackend.service;

import com.example.englishhubbackend.dto.request.ExerciseCreateRequest;
import com.example.englishhubbackend.dto.request.ExerciseUpdateRequest;
import com.example.englishhubbackend.dto.request.QuestionCreateRequest;
import com.example.englishhubbackend.dto.request.QuestionUpdateRequest;
import com.example.englishhubbackend.dto.response.ExerciseResponse;
import com.example.englishhubbackend.dto.response.QuestionGroupResponse;
import com.example.englishhubbackend.dto.response.QuestionResponse;
import java.util.List;
import java.util.UUID;

public interface ExerciseService {
  ExerciseResponse createExercise(UUID courseId, ExerciseCreateRequest exerciseCreateRequest);

  List<ExerciseResponse> getAllExerciseFromCourse(UUID courseID);

  ExerciseResponse updateExercise(UUID exerciseId, ExerciseUpdateRequest exerciseUpdateRequest);

  void deleteExercise(UUID exerciseId);

  List<QuestionResponse> addQuestionsToExercise(
      UUID exerciseId, List<QuestionCreateRequest> questionCreateRequest);

  QuestionResponse addQuestionToExercise(
      UUID exerciseId, QuestionCreateRequest questionCreateRequest);

  List<QuestionResponse> getAllQuestionsFromExercise(UUID exerciseId);

  QuestionResponse updateQuestionInExercise(
      UUID exerciseId, UUID questionId, QuestionUpdateRequest questionUpdateRequest);

  ExerciseResponse getExerciseById(UUID courseId, UUID exerciseId);

  List<QuestionGroupResponse> getQuestionGroupsFromExercise(UUID exerciseId);

  void deleteQuestionFromExercise(UUID exerciseId, UUID questionId);
}
